connections/simopen.md

```diff
@@ -0,0 +1,89 @@
+# Simultaneous Open for bootstrapping connections in multistream-select
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | DRAFT           |
+
+Authors: [@vyzo]
+
+Interest Group: [@raulk], [@stebalien]
+
+[@vyzo]: https://github.com/vyzo
+[@raulk]: https://github.com/raulk
+[@stebalien]: https://github.com/stebalien
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+
+## Introduction
+
+In order to support direct connections through NATs with hole
+punching, we need to account for simultaneous open. In such cases,
+there is no single initiator and responder, but instead both peers act
+as initiators. This breaks protocol negotiation in
+multistream-select, which assumes a single initator.
+
+This draft proposes a simple extension to the multistream protocol
+negotiation in order to select a single initator when both peers are
+acting as such.
+
+## The Protocol
+
+When a peer acting as the initiator enters protocol negotiation, it
+sends the string `iamclient` as first protocol selector. If the other
+peers is a responder or doesn't support the extension, then it
+responds with `na` and protocol negotiation continues as normal.
+
+If both peers believe they are the initiator, then they both send
+`iamclient`. If this is the case, they enter an initiator selection
+phase, where one of the peers is selected to act as the initiator. In
+order to do so, they both generate a random 256-bit integer and send
+it as response to the `iamclient` directive, prefixed with the
+`select:` string. The integer is in big-endian format, encoded in base64.
```

Comment:
@vyzo Why do we need to encode to base64 here ?

Reply:
I'd assume it's because multistream 1 is a text-based, human-readable protocol (with all the pros and cons that come with it).

That said, base64 seems like a suboptimal choice here. I propose switching to hexadecimal encoding. Encoding `math.MaxUint64` (18446744073709551615) is `0xFFFFFFFFFFFFFFFF` in hex and `MTg0NDY3NDQwNzM3MDk1NTE2MTU=` in base64.
Also, 256 bits seem excessive here. I think we can live with a collision probability of 2^-64, which would allow implementations to use a regular `uint64` instead of `math/big.Int`. We might even want to limit this to 52 (?) bits to make things easier in JavaScript (but I'm no expert on that).

Reply:
> Encoding math.MaxUint64 (18446744073709551615) is 0xFFFFFFFFFFFFFFFF in hex and MTg0NDY3NDQwNzM3MDk1NTE2MTU= in base64.

[I think it would be `//////////8=`](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=ba311359861b6ab7459cc0345f57567b)

Reply:
Would it? I used https://www.base64encode.net/, and that gives the same result (except for the last character?) as `echo 18446744073709551615 | base64`. But maybe that's decoding the string "18446744073709551615" and not the number?

As `/` is a separator we use in multistream, so base64 encoding might get us into trouble, if we don't escape it (one of the perks of using a text-based protocol...).

Reply:
Yes I guess you're encoding the string.
It doesn't make sense that encoding something in base64 would give something longer than encoding that same thing in base16

I think that `/` isn't a problem because of the `select:` prefix. Otherwise I'd suggest base58.

Reply:
Not opposed to base58, but not sure if that's easily accessible in all the languages we care about. In Go, we'd have to import a third-party library (it's not in the standard library).

Any thoughts on the length of this value?